src/fastpath/fastpath.c

```diff
@@ -502,7 +731,52 @@ void NORETURN fastpath_reply_recv(word_t cptr, word_t msgInfo)
     callerSlot->cteMDBNode = nullMDBNode;
 #endif
 
-    /* I know there's no fault, so straight to the transfer. */
+#ifdef CONFIG_EXCEPTION_FASTPATH
+    if (unlikely(fault_type != seL4_Fault_NullFault)) {
+        bool_t restart = 0;
+        switch (fault_type) {
+        /* Add cases for other types of faults here */
+        default:
+            restart = 1;
+        }
+
+        /* Note - this works as is for VM faults but will need to be changed when other faults are added. This is because
+         * vm faults always restart the faulting thread upon reply but this is not necessarily the case with other types
+         * of faults, which may instead become inactive. This can either be handled in the slowpath or redirected to the
+         * slowpath, but either way it must be changed so that we do not try and forcefully switch to a thread which is
+         * meant to be inactive. */
+        if (restart) {
+            word_t pc = getRestartPC(caller);
+            setNextPC(caller, pc);
+        }
```

Comment:
This should probably be simplified to just:
```suggestion
        /* Note - this works as is for VM faults but will need to be changed when other faults are added. This is because
         * vm faults always restart the faulting thread upon reply but this is not necessarily the case with other types
         * of faults, which may instead become inactive. This can either be handled in the slowpath or redirected to the
         * slowpath, but either way it must be changed so that we do not try and forcefully switch to a thread which is
         * meant to be inactive. */
        word_t pc = getRestartPC(caller);
        setNextPC(caller, pc);
```